node/utils.go

```diff
@@ -0,0 +1,100 @@
+package node
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"io/ioutil"
+	"strings"
+	"time"
+
+	grpc_retry "github.com/grpc-ecosystem/go-grpc-middleware/retry"
+	"github.com/prometheus/common/log"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/keepalive"
+)
+
+// splitAndTrimEmpty slices s into all subslices separated by sep and returns a
+// slice of the string s with all leading and trailing Unicode code points
+// contained in cutset removed. If sep is empty, SplitAndTrim splits after each
+// UTF-8 sequence. First part is equivalent to strings.SplitN with a count of
+// -1.  also filter out empty strings, only return non-empty strings.
+func splitAndTrimEmpty(s, sep, cutset string) []string {
+	if s == "" {
+		return []string{}
+	}
+
+	spl := strings.Split(s, sep)
+	nonEmptyStrings := make([]string, 0, len(spl))
+	for i := 0; i < len(spl); i++ {
+		element := strings.Trim(spl[i], cutset)
+		if element != "" {
+			nonEmptyStrings = append(nonEmptyStrings, element)
+		}
+	}
+	return nonEmptyStrings
+}
+
+// ConstructDialOptions constructs a list of grpc dial options
+func ConstructDialOptions(
+	extraOpts ...grpc.DialOption,
+) []grpc.DialOption {
+	const (
+		retries            = 50 // 50 * 100ms = 5s total
+		timeout            = 100 * time.Millisecond
```

Comment:
This seems like a very aggressive timeout. Maybe increase a bit? We don't want to hammer the server if it's already struggling. Although I suppose the backoff will mitigate it, and I guess most calls are cheap anyway.

Reply:
With exponential back off this would this timeout would be 1.6s on the 5th attempt. Will increase it 1 second.